reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java

```diff
@@ -5,311 +5,236 @@
+	final long          ttlMillis;
 
-	volatile int                                                   remainingDeferredTasks;
-	static final AtomicIntegerFieldUpdater<BoundedElasticScheduler>REMAINING_DEFERRED_TASKS =
-			AtomicIntegerFieldUpdater.newUpdater(BoundedElasticScheduler.class, "remainingDeferredTasks");
+	volatile BoundedServices boundedServices;
+	static final AtomicReferenceFieldUpdater<BoundedElasticScheduler, BoundedServices> BOUNDED_SERVICES =
+			AtomicReferenceFieldUpdater.newUpdater(BoundedElasticScheduler.class, BoundedServices.class, "boundedServices");
 
+	volatile ScheduledExecutorService evictor;
+	static final AtomicReferenceFieldUpdater<BoundedElasticScheduler, ScheduledExecutorService> EVICTOR =
+			AtomicReferenceFieldUpdater.newUpdater(BoundedElasticScheduler.class, ScheduledExecutorService.class, "evictor");
 
-	BoundedElasticScheduler(int threadCap, int deferredTaskCap, ThreadFactory factory, int ttlSeconds) {
-		if (ttlSeconds < 0) {
-			throw new IllegalArgumentException("ttlSeconds must be positive, was: " + ttlSeconds);
+	/**
+	 * This constructor lets define millisecond-grained TTLs and a custome {@link Clock},
+	 * which can be useful for tests.
+	 */
+	BoundedElasticScheduler(int maxThreads, int maxTaskQueuedPerThread,
+			ThreadFactory threadFactory, long ttlMillis, Clock clock) {
+		if (ttlMillis <= 0) {
+			throw new IllegalArgumentException("TTL must be strictly positive, was " + ttlMillis + "ms");
 		}
-		this.ttlSeconds = ttlSeconds;
-		if (threadCap < 1) {
-			throw new IllegalArgumentException("threadCap must be strictly positive, was: " + threadCap);
+		if (maxThreads <= 0) {
+			throw new IllegalArgumentException("maxThreads must be strictly positive, was " + maxThreads);
 		}
-		if (deferredTaskCap < 1) {
-			throw new IllegalArgumentException("deferredTaskCap must be strictly positive, was: " + deferredTaskCap);
+		if (maxTaskQueuedPerThread <= 0) {
+			throw new IllegalArgumentException("maxTaskQueuedPerThread must be strictly positive, was " + maxTaskQueuedPerThread);
 		}
-		this.threadCap = threadCap;
-		this.remainingThreads = threadCap;
-		this.deferredTaskCap = deferredTaskCap;
-		this.remainingDeferredTasks = deferredTaskCap;
-		this.factory = factory;
-		this.idleServicesWithExpiry = new ConcurrentLinkedDeque<>();
-		this.deferredFacades = new ConcurrentLinkedQueue<>();
-		this.allServices = new ConcurrentLinkedQueue<>();
+		this.maxThreads = maxThreads;
+		this.maxTaskQueuedPerThread = maxTaskQueuedPerThread;
+		this.factory = threadFactory;
+		this.clock = Objects.requireNonNull(clock, "A Clock must be provided");
+		this.ttlMillis = ttlMillis;
+
+		this.boundedServices = new BoundedServices(this);
 		this.evictor = Executors.newScheduledThreadPool(1, EVICTOR_FACTORY);
-		this.evictor.scheduleAtFixedRate(() -> this.eviction(System::currentTimeMillis),
-				ttlSeconds,
-				ttlSeconds,
-				TimeUnit.SECONDS);
+		evictor.scheduleAtFixedRate(boundedServices::eviction,
```

Comment:
nit: `this.boundedServices = `, `this.evictor = ` and `scheduleAtFixedRate` can be replaced with a single `start()` since they share the implementation.

Reply:
they don't really share the implementation since the `BOUNDED_SERVICE` would need to be initialized to `SHUTDOWN` first, and in constructor the looping is superfluous.